common/lib/dependabot/pull_request_creator/message_builder/metadata_presenter.rb

```diff
@@ -0,0 +1,283 @@
+
+        def_delegators :metadata_finder,
+                       :changelog_url,
+                       :changelog_text,
+                       :commits_url,
+                       :commits,
+                       :maintainer_changes,
+                       :releases_url,
+                       :releases_text,
+                       :source_url,
+                       :upgrade_guide_url,
+                       :upgrade_guide_text
+
+        def initialize(dependency:, source:, metadata_finder:,
+                       vulnerabilities_fixed:, github_redirection_service:)
+          @dependency = dependency
+          @source = source
+          @metadata_finder = metadata_finder
+          @vulnerabilities_fixed = vulnerabilities_fixed
+          @github_redirection_service = github_redirection_service
+        end
+
+        def to_s
+          msg = ""
+          msg += vulnerabilities_cascade
+          msg += release_cascade
+          msg += changelog_cascade
+          msg += upgrade_guide_cascade
+          msg += commits_cascade
+          msg += maintainer_changes_cascade
+          msg += break_tag unless msg == ""
+          "\n" + sanitize_links_and_mentions(msg, unsafe: true)
+        end
+
+        private
+
+        def vulnerabilities_cascade
+          return "" unless vulnerabilities_fixed&.any?
+
+          msg = ""
+          vulnerabilities_fixed.each do |v|
+            msg += serialized_vulnerability_details(v)
+          end
+
+          msg = sanitize_template_tags(msg)
+          msg = sanitize_links_and_mentions(msg)
+
+          build_details_tag(summary: "Vulnerabilities fixed", body: msg)
+        end
+
+        def release_cascade
+          return "" unless releases_text && releases_url
+
+          msg = "*Sourced from [#{dependency.display_name}'s releases]"\
+                "(#{releases_url}).*\n\n"
+          msg +=
+            begin
+              release_note_lines = releases_text.split("\n").first(50)
+              release_note_lines =
+                release_note_lines.map { |line| "> #{line}\n" }
+              if release_note_lines.count == 50
+                release_note_lines << truncated_line
+              end
+              release_note_lines.join
+            end
+          msg = link_issues(text: msg, dependency: dependency)
+          msg = fix_relative_links(
+            text: msg,
+            base_url: source_url + "/blob/HEAD/"
+          )
+          msg = sanitize_template_tags(msg)
+          msg = sanitize_links_and_mentions(msg)
+
+          build_details_tag(summary: "Release notes", body: msg)
+        end
+
+        def changelog_cascade
+          return "" unless changelog_url && changelog_text
+
+          msg = "*Sourced from "\
+                "[#{dependency.display_name}'s changelog]"\
+                "(#{changelog_url}).*\n\n"
+          msg +=
+            begin
+              changelog_lines = changelog_text.split("\n").first(50)
```

Comment:
Idea: `truncate` helper? 
This seems like `release_note_lines` above and `upgrade_lines` below meeting rule of threes. Meh as existing code.

Reply:
See: https://github.com/dependabot/dependabot-core/pull/2540